File: Data Structures & Algorithms/count-number-of-islands/submission-6.java
class Solution {
    public int numIslands(char[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        int count = 0;
        boolean[][] visited = new boolean[rows][cols];
        for(int r = 0; r < rows; r++){
            for(int c = 0; c < cols; c++){
                if(grid[r][c] == '1' && !visited[r][c]){
                    dfs(r, c, grid, visited);
                    count++;
                    
                }
            }
        }
        return count;
        
    }
    private void dfs(int r, int c, char[][] grid, boolean[][] visited){
        //boundary checks
        if(r < 0 || r >= grid.length || c < 0 || c >= grid[0].length){
            return;
        }
        

        //if water or we visited the coordinate return
        if(grid[r][c] == '0' || visited[r][c]) return;


        visited[r][c] = true;
        dfs(r + 1, c, grid, visited);
        dfs(r - 1, c, grid, visited);
        dfs(r, c + 1, grid, visited);
        dfs(r, c - 1, grid, visited);

    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-6.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        '''
            Given a grid, grid[i] = 0 (water), 1 = land

            island is defined as a group of 1s connected horizontally or vertically

            area of an island is number of cells wwithin the island

            DFS in 4 dirs add up area and return that p calm
        '''
        ROWS, COLS = len(grid), len(grid[0])
        visited = set()
        area = 0
        def dfs(r,c):
            if r < 0 or c < 0 or r >= ROWS or c >= COLS:
                return 0
            if grid[r][c] == 0 or (r,c) in visited:
                return 0
            visited.add((r,c))
            area = (1 +
                dfs(r + 1, c) +
                dfs(r, c + 1) +
                dfs(r - 1, c) +
                dfs(r, c - 1)
            )
            return area

        for r in range(ROWS):
            for c in range(COLS):
                area = max(area, dfs(r,c))
        return area
                

        